src/hotspot/share/cds/archiveBuilder.cpp

```diff
@@ -1037,44 +1039,162 @@ class ArchiveBuilder::CDSMapLogger : AllStatic {
+      _heap_info(heap_info), _st(st), _source_obj(src_obj) {}
+    void do_field(fieldDescriptor* fd) {
+      _st->print(" - ");
+      BasicType ft = fd->field_type();
+      switch (ft) {
+      case T_ARRAY:
+      case T_OBJECT:
+        fd->print_on(_st); // print just the name and offset
+        print_oop_with_requested_addr_cr(_st, _source_obj->obj_field(fd->offset()));
+        break;
+      default:
+        if (ArchiveHeapWriter::is_marked_as_native_pointer(_heap_info, _source_obj, fd->offset())) {
+          LP64_ONLY(assert(ft == T_LONG, "must be"));
+          NOT_LP64 (assert(ft == T_INT, "must be"));
+
+          // We have a field that looks like an integer, but it's actually a pointer to a MetaspaceObj.
+          address source_native_ptr = (address)
+            LP64_ONLY(_source_obj->long_field(fd->offset()))
+            NOT_LP64( _source_obj->int_field (fd->offset()));
+          ArchiveBuilder* builder = ArchiveBuilder::current();
+
+          // The value of the native pointer at runtime.
+          address requested_native_ptr = builder->to_requested(builder->get_buffered_addr(source_native_ptr));
+
+          // The address of _source_obj at runtime
+          oop requested_obj = ArchiveHeapWriter::source_obj_to_requested_obj(_source_obj);
+          // The address of this field in the requested space
+          address requested_field_addr = cast_from_oop<address>(requested_obj) + fd->offset();
+
+          fd->print_on(_st);
+          _st->print_cr(PTR_FORMAT " (marked metadata pointer @" PTR_FORMAT " )",
+                        p2i(requested_native_ptr), p2i(requested_field_addr));
+        } else {
+          fd->print_on_for(_st, _source_obj); // name, offset, value
+          _st->cr();
+        }
+      }
+    }
+  };
+
+  // Print the fields of instanceOops, or the elements of arrayOops
+  static void log_oop_details(ArchiveHeapInfo* heap_info, oop source_oop) {
+    LogStreamHandle(Trace, cds, map, oops) st;
+    if (st.is_enabled()) {
+      Klass* source_klass = source_oop->klass();
+      ArchiveBuilder* builder = ArchiveBuilder::current();
+      Klass* requested_klass = builder->to_requested(builder->get_buffered_addr(source_klass));
+
+      st.print(" - klass: ");
+      source_klass->print_value_on(&st);
+      st.print(" " PTR_FORMAT, p2i(requested_klass));
+      st.cr();
+
+      if (source_oop->is_typeArray()) {
+        TypeArrayKlass::cast(source_klass)->oop_print_elements_on(typeArrayOop(source_oop), &st);
+      } else if (source_oop->is_objArray()) {
+        objArrayOop source_obj_array = objArrayOop(source_oop);
+        for (int i = 0; i < source_obj_array->length(); i++) {
+          st.print(" -%4d: ", i);
+          print_oop_with_requested_addr_cr(&st, source_obj_array->obj_at(i));
+        }
+      } else {
+        st.print_cr(" - ---- fields (total size " SIZE_FORMAT " words):", source_oop->size());
```

Comment:
Nit, proposal: "fields (xx words):" (no leading dashes, no "total size") for more density